### DISCUSSION TITLE
Support setting Cron in Pipeline

### DISCUSSION THREAD

**Discussion by JakobDev** [woodpecker-ci/woodpecker#2283] [Request a Feature]
### Clear and concise description of the problem

If you want to setup a Cronjob, you have to do it in the UI and add a cron event filter to the Pipelne. It would be great, if Cron can be used directly in the Pipeline.

### Suggested solution

Something like this:
```yaml
when:
    event: cron
    cron: some-name
    schedule: @daily
```

### Additional context

This would make it easier to copy and paste Pipelines in other Projects. I also think, that as much as possible should be configured with the Pipeline files and not with the GUI.

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by 6543** [comment]
we would somehow get that entry into DB ... and maintain it acordingly

To Decide: how to determine the reference to run on and can there ocure config conflicts between branches ..

yes you could have multible crons with same name specifyed and so it would not be unique :bomb:

**Reply by Janik-Haag** [reply]
Would adding a extra section named cron like with clone and steps be a possibility and then diffing that section with the database entry on every commit/push to main? Then you could also make the pipeline fail when there is multiple cron jobs with the same name and keep the old entry in the database until it is fixed.

Btw would PRs for this be accepted, I want to start dabbling with go a bit more anyways but can't promise anything.

**Reply by qwerty287** [reply]
> Would adding a extra section named cron like with clone and steps be a possibility and then diffing that section with the database entry on every commit/push to main? Then you could also make the pipeline fail when there is multiple cron jobs with the same name and keep the old entry in the database until it is fixed.

Yes, that's the only idea we had yet I think. However, the issue is: What happens if you have multiple branches? Just doing the update for `main` will prevent other branches from being able to run crons which still should be possible.

> Btw would PRs for this be accepted

Of course!

**Reply by Janik-Haag** [reply]
> Yes, that's the only idea we had yet I think. However, the issue is: What happens if you have multiple branches? Just doing the update for main will prevent other branches from being able to run crons which still should be possible.

hm what if you have a syntax like:
```yml
cron:
  renovateUpdates:
    branch: develop
    schedule: @daily
  ...
```
and only apply it to the database when it changes in the default branch? Changing it on a per branch basis does seem a bit mad since you never know the current state without checking the web interface or every single branch. 
If we go with the small example above we could also assume that if branch is not set/a empty string that the user means the default branch, because git allows almost any char in branch names except for `\` from my quick testing.

**Reply by qwerty287** [reply]
Yes, this looks like a good solution 👍

**Reply by Janik-Haag** [reply]
Then we would still need to figure out how to do migrations, because just wiping the existing cronjobs with the declarative ones seems like it could make a few users quite unhappy but that's probably a implementation detail to worry about later.

**Reply by qwerty287** [reply]
I think updating the yaml should delete all crons that were previously defined but removed then. What would you suggest otherwise?

If you mean crons defined via ui: I'd still support both ways, at least for now. You could add a new boolean db column which is used to check whether it's yaml or ui defined.

Or do you mean something completely different?

**Reply by Janik-Haag** [reply]
Im fine with just deleting, but I thought about maybe require the user to maybe have to acknowledge that the current crons will be deleted.

**Reply by Janik-Haag** [reply]
I should probably add a file like `pipeline/frontend/yaml/types/cron.go` and just add a struct with stuff like branch, timer, etc.. 
How does the data flow from .woodpecker file/folder in the git repo to getting parsed to writing stuff to the database work (I read the architecture diagram but couldn't make much sense out of it)? 
While looking through the code I also found:
```go
	// Cron is the JSON data of a cron job
	Cron struct {
		ID        int64  `json:"id"`
		Name      string `json:"name"`
		RepoID    int64  `json:"repo_id"`
		CreatorID int64  `json:"creator_id"`
		NextExec  int64  `json:"next_exec"`
		Schedule  string `json:"schedule"`
		Created   int64  `json:"created_at"`
		Branch    string `json:"branch"`
	}
```
Name, Schedule, and Branch are obvious. we probably also can get the repo_id from somewhere:tm: and use the git commit time as created timestap, we are probably able to leave next_exec untouched, but I don't know what to do about id and creatore_id. we could probably make creator_id a optional value because git blame would tell you who modified the cron job.

**Reply by qwerty287** [reply]
This struct isn't the db model I think(it's here: https://github.com/woodpecker-ci/woodpecker/blob/main/server/model/cron.go). Yes, I think you can skip the creator. The ID field is automatically set when you insert it into the db, you don't need to care about this.
